### ISSUE TITLE
Refactor distributions to objects

### ISSUE THREAD

**Issue by balping** [markrogoyski/math-php#247]
Currently distributions are static interfaces that follow an approach borrowed by functional languages. Although you can use PHP in a functional style, it is clearly object oriented and in the case of distributions, we could use the benefits of OOP.

An instance of a distribution could be a distribution function with  given parameters. And by 'function' I mean the whole abstract pdf/cdf functions, without evaluated for a certain value. Probably the easiest way to explain this is via an example. So take this example of the binomial distribution:

```php
// current
$v1 = Binomial::pmf($n, $r, $P);
$v2 = Binomial::cdf($n, $r, $P);
```

would be changed to this:

```php
// planned
$binom = new Binomial($r, $P);
// we defined a distribution object with given parameters, now we can calculate things
$v1 = $binom->pmf($n);
$v2 = $binom->cdf($n);
```

This approach  would have the following benefits:
* separating independent variables from parameters
* a consistent api across different distributions, maybe forced by an interface
* consistency with categorical distribution that works in an object oriented way already
* possibility to validate parameters at initialisation and not at computation (eg. now in recursive cdf functions the same parameters are validated many times unnecessarily)
* possibility to add any number of helper methods (eg. min/max of independent variable, see below)
* it would make it _much_ easier and more efficient to implement random number generation on discrete distributions:
  -  min/max values of the independent variable (_0_ and _r_ for binomial) 
 could be queried via a simple method call
  - lookup tables could be cached if we need multiple random values
  - the api could be as simple as this: `$rand = $binom->random();`

cons:
* heavily breaking change, but this library is pre 1.0, so I wouldn't consider this as a serious issue

I'm absolutely willing to do this refactoring (and the implementation of random on discrete distributions)  but I wouldn't  start working on it without the approval of the maintainers, because I wouldn't like to waste my time.

**Comment by markrogoyski** [comment]
Hi @balping 
Thank you for your interest in MathPHP and making the library better.

This has been suggested before, and I think it makes sense.

Generally, if something is deterministic, depending only on its input, then we usually make it static. For example:
```php
Algebra::gcd(8, 12)
```

If something has state, or if multiple things can be computed from some initial values, then it is an object with methods. For example, Matrix, Vector, Regressions, etc.

In the case of probability distributions, the original thinking was that the PDF or CDF are deterministic based on all the input parameters. However, you could think of it as a distribution is made up of parameters, and then using those parameters, you can compute the PDF or CDF using different support values.

As you mention, the library is pre 1.0, so this is the time to make drastic backwards-incompatible changes if ever.

My only concern is if we were to make this change, to not go overboard. It should still be simple and straightforward to do math operations. Most math packages like R and NumPy generally are function  based and both R and NumPy are very popular with mathematicians. Although, those languages have a stronger command-line interpreter culture where you want concise functions. 

We would want a nice balance between clean, DI-friendly object-oriented code, and at the same time not have the library get in the way of actually computing things (you know, doing math!). Your proposal seems like it would strike that balance.

I would like to get input from others though. Let's leave this issue open for at least a week or so to give others that may be interested time to consider and comment.

@Beakerboy What are you thoughts on this?

Thanks again for making the suggestion.

**Comment by Beakerboy** [comment]
I think moving more towards OOP in the distribution namespace is the right move.

**Comment by balping** [comment]
There's a way of moving to OOP _and_ keeping an optional use of static interface by creating aliases like this:

```php
public static function cdf($var, ... $params){
   return (new static(...$params)->cdf($var);
}
```

In my opinion this kind of aliasing would just confuse people, but it's certainly a way of having both objects and deterministic static functions.

**Comment by Beakerboy** [comment]
I think we should have one way to do things, object oriented. We should make an interface to the abstract parent class. The Distribution class is empty, so if there is any food nationality we can combine for both continuous and discrete distributions, we should consider that.

**Comment by markrogoyski** [comment]
I've created a branch _distribution_ for this project so multiple people can work on it and give visibility to what has been done and what is remaining. Please point pull requests to this branch.

For each distribution that is not currently object based:
- Create a constructor that takes the distribution parameters as constructor parameters
- Move distribution parameter limit checks into the constructor
- Change all the methods from static to object methods
- Change the PDF/PMF methods to only take the support values as method parameters
- Change the CDF methods to only take support values as method parameters
- Update unit tests
- Improve unit tests (Add docBlocks, testCase descriptions, etc.) to match latest standard

All distribution namespaces to be refactored:
- Continuous
- Discrete
- Multivariate

I think we can consider refactoring base classes and or interfaces after the work is done as a new issue as to keep the scope from getting too large and allow the unit tests to serve as full regression tests for this task.

**Comment by Beakerboy** [comment]
I was thinking about adding something like this to the parents continuous class 
    
        public function __construct(...$params)
    {
        $limits = static::LIMITS;
        array_shift($limits);
        $new_params = $limits;
        $number_needed = count($limits);
        $number_given = count($params);
        if ($number_needed !== $number_given) {
            throw new Exception\BadDataException('Incorrect number of parameters. Need ' . $number_needed . ' but given ' . $number_given . '.');
        } else {
            $i = 0;
            foreach ($limits as $key => $value) {
                $new_params[$key] = $value[$i++]; 
            }
            Support::checkLimits($limits, $new_params);
        }
        $this->params = $new_params;
        print_r($new_params);
    }

**Comment by markrogoyski** [comment]
Since there are multiple people working on this, make sure you git pull the latest version of the distribution branch before starting work each time to make sure you don't work on something that has already been done.

**Comment by Beakerboy** [comment]
will do. thanks for the reminder

**Comment by markrogoyski** [comment]
This has been fully implemented in [v0.37.0](https://github.com/markrogoyski/math-php/releases/tag/v0.37.0).
